axum-extra/src/response/error_response.rs

```diff
@@ -0,0 +1,59 @@
+use axum_core::response::{IntoResponse, Response};
+use http::StatusCode;
+use std::error::Error;
+use std::io::Write;
+use tracing::error;
+
+/// Convenience response to create an error response from a non-IntoResponse error
+///
+/// This provides a method to quickly respond with an error that does not implement
+/// the IntoResponse trait itself. This type should only be used for debugging purposes or internal
+/// facing applications, as it includes the full error chain with descriptions,
+/// thus leaking information that could possibly be sensitive.
+///
+/// ```rust
+/// use axum_extra::response::InternalServerError;
+/// use axum_core::response::IntoResponse;
+/// # use std::io::{Error, ErrorKind};
+/// # fn try_thing() -> Result<(), Error> {
+/// #   Err(Error::new(ErrorKind::Other, "error"))
+/// # }
+///
+/// async fn maybe_error() -> Result<String, InternalServerError<Error>> {
+///     try_thing().map_err(InternalServerError)?;
+///     // do something on success
+///     # Ok(String::from("ok"))
+/// }
+/// ```
+#[derive(Debug)]
+pub struct InternalServerError<T>(pub T);
+
+impl<T: Error> IntoResponse for InternalServerError<T> {
+    fn into_response(self) -> Response {
+        let mut error: Vec<String> = Vec::new();
+        error.push(format!("{}", self.0));
+        let mut e: &dyn Error = &self.0;
+        while let Some(new_e) = e.source() {
+            e = new_e;
+            error.push(format!(": {e}"))
+        }
+        error!("Internal server error: {}", error);
```

Comment:
Tracing has built-in support for `dyn Error`s as values you can capture as fields. I think it would be good to ude that here and add an `error` field, instead of string-interpolating the error into the message.

Reply:
Thats very nice, I did not know that.

Reply:
The `%` you added bypasses the dyn Error support.. Did it not compile previously?

Reply:
Ah okay, was not aware of that. I did it that way because with `error!(error = error)`, the compiler complains that a bound is not satisfied:

```
error[E0277]: the trait bound `T: tracing::Value` is not satisfied
  --> axum-extra/src/response/error_response.rs:33:9
   |
33 |         error!(error = err);
   |         ^^^^^^^^^^^^^^^^^^^ the trait `tracing::Value` is not implemented for `T`, which is required by `&T: tracing::Value`
   |
   = note: required for `&T` to implement `tracing::Value`
   = note: required for the cast from `&&T` to `&dyn tracing::Value`
   = note: this error originates in the macro `$crate::valueset` which comes from the expansion of the macro `error` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I googled about this and found the approach with `%`

Reply:
You need to cast to `&dyn Error`. This compiles: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b08c7457dc0862f279b3b96daa7fca24

Reply:
That fixes the trait bound, but it then asks me to add  a  'static to T

Reply:
Fixed, the `tracing::Value` impl for `dyn Error` was simply behind a feature that wasn't always getting enabled.